RFC: project goals for 2025h1

nikomatsakis opened the pull request:
Propose a slate of 40 project goals for 2025H1, including 3 flagship goals:

* Continue making Rust easier to use for network systems by [**bringing the Async Rust experience closer to parity with sync Rust**](https://rust-lang.github.io/rust-project-goals/2025h1/./async.html). In 2025H1 we plan to:
    * tell a complete story for the use of async fn in traits, unblocking wide ecosystem adoption;
    * improve the ergonomics of `Pin`, which is frequently used in low-level async code; and
    * prepare to support asynchronous (and synchronous) generators in the language.
* Continue helping Rust support low-level projects by [**stabilizing compiler options and tooling used by the Rust-for-Linux project**](https://rust-lang.github.io/rust-project-goals/2025h1/./rfl.html). In 2025H1 we plan to:
    * implement [RFC #3716](https://github.com/rust-lang/rfcs/pull/3716) to allow stabilizing ABI-modifying compiler flags to control code generation, sanitizer integration, and so forth;
    * taking the first step towards stabilizing [`build-std`](https://rust-lang.github.io/rust-project-goals/2025h1/https://doc.rust-lang.org/cargo/reference/unstable.html#build-std) by [creating a stable way to rebuild core with specific compiler options](./build-std.html);
    * add rustdoc features to extract and customize rustdoc tests (`--extract-doctests`);
    * stabilize clippy configuration like `.clippy.toml` and `CLIPPY_CONF_DIR`;
    * stabilize compiler flags to extract dependency info (e.g., as via `-Zbinary-dep-depinfo=y`) and to configure no-std without requiring it in the source file (e.g., as via `-Zcrate-attr`);
* Address the biggest concerns raised by Rust maintainers, lack of face-to-face interaction, by [**organizing the Rust All-Hands 2025**](https://rust-lang.github.io/rust-project-goals/2025h1/./all-hands.html). In 2025H1 we plan to:
    * convene Rust maintainers to celebrate Rust's tenth birthday at [RustWeek 2025](https://2025.rustweek.org) (co-organized with [RustNL](https://2025.rustweek.org/about/);
    * author a first draft for a [Rust vision doc](https://rust-lang.github.io/rust-project-goals/2025h1/./rust-vision-doc.html) and gather feedback.


[Rendered](https://github.com/nikomatsakis/rfcs/blob/2025h1/text/3764-Project-Goals-2025h1.md)

programmerjake commented on text/0000-Project-Goals-2025h1.md:
```diff
@@ -0,0 +1,653 @@
+    * prepare to support asynchronous (and synchronous) generators in the language.
+* Continue helping Rust support low-level projects by [**stabilizing compiler options and tooling used by the Rust-for-Linux project**](https://rust-lang.github.io/rust-project-goals/2025h1/./rfl.html). In 2025H1 we plan to:
+    * implement [RFC #3716](https://github.com/rust-lang/rfcs/pull/3716) to allow stabilizing ABI-modifying compiler flags to control code generation, sanitizer integration, and so forth;
+    * taking the first step towards stabilizing [`build-std`](https://rust-lang.github.io/rust-project-goals/2025h1/https://doc.rust-lang.org/cargo/reference/unstable.html#build-std) by [creating a stable way to rebuild core with specific compiler options](./build-std.html);
```
`./build-std.html` is a broken link

AustinWise replied:
```suggestion
    * taking the first step towards stabilizing [`build-std`](https://doc.rust-lang.org/cargo/reference/unstable.html#build-std) by [creating a stable way to rebuild core with specific compiler options](./build-std.html);
```

This link was broken.

AustinWise commented on text/0000-Project-Goals-2025h1.md:
```diff
@@ -0,0 +1,653 @@
+| ["Stabilizable" prototype for expanded const generics](https://rust-lang.github.io/rust-project-goals/2025h1/min_generic_const_arguments.html)                      | [Boxy][]         | [lang], [types]                                     |
+| [Bring the Async Rust experience closer to parity with sync Rust](https://rust-lang.github.io/rust-project-goals/2025h1/async.html)                                 | [Tyler Mandry][]         | [compiler], [lang], [libs-api], [types]             |
+| [Continue resolving `cargo-semver-checks` blockers for merging into cargo](https://rust-lang.github.io/rust-project-goals/2025h1/cargo-semver-checks.html)          | [Predrag Gruevski][]      | [cargo], [rustdoc]                                  |
+| [Declarative (https://rust-lang.github.io/rust-project-goals/2025h1/`macro_rules!`) macro improvements](macro-improvements.html)                                    | [Josh Triplett][]    | [lang], [wg-macros]                                 |
```
```suggestion
| [Declarative (`macro_rules!`) macro improvements](macro-improvements.html)                                    | [Josh Triplett][]    | [lang], [wg-macros]                                 |
```

This had an extra URL added in the title, I updated to match the title of the [linked document](https://rust-lang.github.io/rust-project-goals/2025h1/macro-improvements.html).

nikomatsakis replied:
Thanks. There's something bogus in the script that needs to be fixed.

clarfonthey commented:
Given that the goals for 2024H2 had an ["unusual ratification procedure"](https://github.com/rust-lang/rfcs/pull/3672), I assume that the same applies here?

Adding a bit onto this: I'm not 100% sure why this is shared in the form of an RFC, since as the RFC clarifies, a lot of these goals boil down to someone actually having the resources to accomplish them, either in the form of money, time, etc. So, it feels unusual to open an RFC on the specific goals that have been declared, since it feels like the only options for discussion are bad: either criticising people on what they're doing with their time, or asking people to work on things that they don't have the resources for.

I'm not sure what a better alternative forum would be for announcing these goals, asking people to get involved, etc., but I also don't think that an RFC with everything in one giant list is a good solution. Perhaps it could be done in a separate "goals" repo with individual discussion threads for individual goals, or in a similar category on Internals. I personally dislike using Zulip for longer-form discussions, which is why I specifically mention making a repo or Internals category for this.

For the remainder of my feedback, I think that it's probably more appropriate to not comment on the specific goals, but rather, the process surrounding them, which I guess is probably the real goal of this RFC. And so, I'll add my concerns in individual sections in this comment.

---

What steps are done to ensure that these goals are resilient to changing circumstances and burnout? Imagine that a hypothetical person is being paid by their employer to perform a given task, and then they later get laid off? In this circumstance, I would imagine that the answer would be unequivocally that this goal is now off the table, since this person should not feel pressured to work on a goal in their spare time when their source of income and time is immediately severed. However, I have a feeling that many people in this situation would feel pressured to continue working on their goal despite the circumstances, and I think that there should be something in the process to help ensure that people get the support they need in these situations.

And, this is a bit picky, but I think that the term "Goals" itself is putting expectations in the wrong place. As I mentioned, if someone's circumstances change and they can't complete the goal (which is very likely to happen for at least a few goals given that we're committing to *six months* in the *tech industry*), they will feel pressured to complete it. I think that calling this "project plans" or something similar would be worthwhile. I don't really care what colour the shed is painted, and as far as I'm concerned, it can be just decided, but I really don't like the idea of pressuring volunteers into doing something even if their circumstances change.

I especially thing that there should be a culture of acceptance and forgiveness when it comes to burnout, and adding deadlines onto something that may be entirely volunteer work is definitely a way to make it more stressful. I don't have a good solution to this: I know that many people who work on these teams *do* have the support of their employer to work on the project, but many people also don't. The "exposure" of being even the point of contact for a major project goal is effectively worthless, although I personally know that working on open source work can be rewarding in its own right. I think that there should always be a balance and an escape hatch for people who can't handle the pressure when something isn't their full-time job.

---

At least from an outside perspective, a lot of these goals appear to be spreading the project very thin. The prevalence of "help wanted" across a large number of goals, including one of the stated major goals (Rust for Linux), feels very concerning. I truly doubt that this is the plan, but a lot of the goals give the appearance of someone wanting a feature, deciding to be project manager for that feature, and then sitting back and folding their arms and waiting for people to come and implement the feature. Obviously, as an open source project, a lot of contributions hinge on volunteers, but it feels premature to declare something a "goal" without even deciding the people involved with its implementation.

I think that in the future, anything with the title "goal" should have all of its resources planned ahead of time, especially with contingencies and backups in case of unexpected situations as mentioned above.

---

Similarly, "invited goals" feel like a massive commitment with little benefit other than the potential for recognition, which even then feels unlikely. I feel like the project already has a decent system for something more "goal" like, which is the Rust Foundation grant program: people with dedicated plans to do something are literally given the resources to complete it, rather than just being given a mentor and getting on the hook for completing something they didn't fully understand from the start. Of course, resources are obviously limited in this case, and not a whole lot of grant money can be doled out for things.

I think that, if we're working outside the framework of literally giving people the resources to complete things, these should be reframed in terms of working groups instead: have the capacity reserved on the team for mentors still, but instead of treating it as a goal for someone to take on, treat it as an area where people can enter and incrementally bring it toward the finish line, then leave as they wish. Part of this is the expectation that something may not be worked on at all, and this is okay.

nikomatsakis commented:
@clarfonthey correct, the ratification procedure will be the same as last time, I was just giving it a bit of time for people to point at mistakes or other things. I'll post the comment today.

Regarding the use of an RFC, I am opening it in this form so as to invite comment. RFCs are the Rust project's mechanism for ratifying consensus on a particular direction (and, to a lesser extent, advertising that to people actively following the project). I agree that sometimes the procedure isn't a perfect fit (and maybe this is one of those cases) but it's still the closest we've got. =)

I agree with you that for this particular RFC the primary "audience" are Rust team members (similar to RFCs about team procedure, for example), but that doesn't mean others have nothing to contribute. For example, perhaps you feel the wording or framing of the flagship goals could be improved -- or that one of the project goals might be elevated to a flagship goal? Or, perhaps you realize that some work you were considering contributing is not on this list, and would like to ask whether you can add a goal of your own?

And, yes, there is room for constructive feedback as well. I appreciate your concern about "telling people what to do with their time" and I agree that many negative comments would be inappropriate and pointless. But I think highlighting potential pitfalls, suggesting alternatives that could be better or ought to be mentioned (at least in the FAQ), or even saying "I don't really think this direction is good" is not out of scope.

nikomatsakis commented:
Regarding invited goals and the grant program, the two work together. Invited goals are one way for the Foundation to know what kind of work would be welcome and hence where grants would be well suited.

nikomatsakis commented:
# Proposed Final Comment Period

Due to its cross-team nature, this RFC follows an unusual ratification procedure. To enter FCP, the **leads** of all impacted teams must check their boxes. Other team members are free to check their boxes but not required for FCP. Any team member from an impacted team should feel free to voice concerns.

*Guidance for reviewing:*

* For each team, the RFC contains a list of "team asks" that have been asked of you by various goals. Check that your team is capable of fulfilling these. We don't want to make promises we can't keep, so try to guess if the sum total of the asks will exceed capacity. If you think it will, we can figure out how to resolve that (either by removing or weakening goals).
* For each goal, check that you are aligned with the work over the next 6 months and *loosely* aligned with the shiny future. The shiny future can always change, but we do want to ensure loose alignment because sometimes people won't want to spend the next 6 months working on feature X unless the team is aligned with the larger vision.

What follows is a check-list of names from each team (generated from the teams repo). Note that only the leads are cited using github pings because otherwise we exceed the limit of 50 names. Please check your box.

## bootstrap

* [ ] @Mark-Simulacrum (required, lead)
* [ ] @onur-ozkan (required, lead)
* [ ] albertlarsan68 (optional)
* [ ] clubby789 (optional)
* [ ] jieyouxu (optional)
* [ ] Kobzol (optional)

## cargo

* [ ] @ehuss (required, lead)
* [ ] arlosi (optional)
* [ ] Eh2406 (optional)
* [ ] epage (optional)
* [ ] joshtriplett (optional)
* [ ] Muscraft (optional)
* [ ] Rustin170506 (optional)
* [ ] weihanglo (optional)

## clippy

* [ ] @flip1995 (required, lead)
* [ ] @Manishearth (required, lead)
* [ ] Alexendoo (optional)
* [ ] blyxyas (optional)
* [ ] Centri3 (optional)
* [ ] dswij (optional)
* [ ] Jarcho (optional)
* [ ] llogiq (optional)
* [ ] matthiaskrgr (optional)
* [ ] xFrednet (optional)
* [ ] y21 (optional)

## compiler

* [ ] @davidtwco (required, lead)
* [ ] @wesleywiser (required, lead)
* [ ] alexcrichton (optional)
* [ ] apiraino (optional)
* [ ] b-naber (optional)
* [ ] bjorn3 (optional)
* [ ] BoxyUwU (optional)
* [ ] camelid (optional)
* [ ] chenyukang (optional)
* [ ] ChrisDenton (optional)
* [ ] cjgillot (optional)
* [ ] compiler-errors (optional)
* [ ] cuviper (optional)
* [ ] DianQK (optional)
* [ ] durin42 (optional)
* [ ] eholk (optional)
* [ ] est31 (optional)
* [ ] estebank (optional)
* [ ] fee1-dead (optional)
* [ ] flodiebold (optional)
* [ ] fmease (optional)
* [ ] jackh726 (optional)
* [ ] jieyouxu (optional)
* [ ] jswrenn (optional)
* [ ] lcnr (optional)
* [ ] lqd (optional)
* [ ] lukas-code (optional)
* [ ] m-ou-se (optional)
* [ ] Mark-Simulacrum (optional)
* [ ] matthewjasper (optional)
* [ ] Nadrieril (optional)
* [ ] nagisa (optional)
* [ ] nikic (optional)
* [ ] nikomatsakis (optional)
* [ ] nnethercote (optional)
* [ ] Noratrieb (optional)
* [ ] oli-obk (optional)
* [ ] petrochenkov (optional)
* [ ] RalfJung (optional)
* [ ] rcvalle (optional)
* [ ] saethlin (optional)
* [ ] scottmcm (optional)
* [ ] SparrowLii (optional)
* [ ] spastorino (optional)
* [ ] TaKO8Ki (optional)
* [ ] tgross35 (optional)
* [ ] the8472 (optional)
* [ ] tmandry (optional)
* [ ] tmiasko (optional)
* [ ] Urgau (optional)
* [ ] WaffleLapkin (optional)
* [ ] workingjubilee (optional)
* [ ] Zalathar (optional)

## crates-io

* [ ] @jtgeibel (required, lead)
* [ ] @Turbo87 (required, lead)
* [ ] carols10cents (optional)
* [ ] eth3lbert (optional)
* [ ] LawnGnome (optional)
* [ ] mdtro (optional)
* [ ] Rustin170506 (optional)

## infra

* [ ] @jdno (required, lead)
* [ ] @shepmaster (required, lead)
* [ ] kennytm (optional)
* [ ] Kobzol (optional)
* [ ] marcoieni (optional)
* [ ] Mark-Simulacrum (optional)
* [ ] pietroalbini (optional)

## lang

* [ ] @nikomatsakis (required, lead)
* [ ] @tmandry (required, lead)
* [ ] joshtriplett (optional)
* [ ] scottmcm (optional)
* [ ] traviscross (optional)

## leadership-council

* [ ] eholk (optional)
* [ ] ehuss (optional)
* [ ] jamesmunns (optional)
* [ ] m-ou-se (optional)
* [ ] Mark-Simulacrum (optional)
* [ ] oli-obk (optional)
* [ ] traviscross (optional)

## libs

* [ ] @Amanieu (required, lead)
* [ ] @m-ou-se (required, lead)
* [ ] cuviper (optional)
* [ ] joshtriplett (optional)
* [ ] the8472 (optional)
* [ ] thomcc (optional)

## libs-api

* [ ] Amanieu (optional)
* [ ] BurntSushi (optional)
* [ ] dtolnay (optional)
* [ ] joshtriplett (optional)
* [ ] m-ou-se (optional)

## opsem

* [ ] @JakobDegen (required, lead)
* [ ] @RalfJung (required, lead)
* [ ] CAD97 (optional)
* [ ] digama0 (optional)
* [ ] saethlin (optional)

## project-stable-mir

* [ ] @celinval (required, lead)
* [ ] oli-obk (optional)
* [ ] scottmcm (optional)

## release

* [ ] @Mark-Simulacrum (required, lead)
* [ ] BoxyUwU (optional)
* [ ] cuviper (optional)
* [ ] Dylan-DPC (optional)
* [ ] pietroalbini (optional)
* [ ] tmandry (optional)

## rustdoc

* [ ] @GuillaumeGomez (required, lead)
* [ ] aDotInTheVoid (optional)
* [ ] camelid (optional)
* [ ] fmease (optional)
* [ ] jsha (optional)
* [ ] Manishearth (optional)
* [ ] Nemo157 (optional)
* [ ] notriddle (optional)

## spec

* [ ] @JoelMarcey (required, lead)
* [ ] ehuss (optional)
* [ ] m-ou-se (optional)
* [ ] nikomatsakis (optional)
* [ ] traviscross (optional)

## testing-devex

* [ ] @calebcartwright (required, lead)
* [ ] epage (optional)
* [ ] Muscraft (optional)
* [ ] thomcc (optional)
* [ ] weihanglo (optional)

## types

* [ ] @jackh726 (required, lead)
* [ ] @lcnr (required, lead)
* [ ] aliemjay (optional)
* [ ] BoxyUwU (optional)
* [ ] compiler-errors (optional)
* [ ] nikomatsakis (optional)
* [ ] oli-obk (optional)
* [ ] spastorino (optional)

## wg-macros

* [ ] @eholk (required, lead)
* [ ] @vincenzopalazzo (required, lead)
* [ ] CohenArthur (optional)
* [ ] danielhenrymantilla (optional)
* [ ] jhpratt (optional)
* [ ] tgross35 (optional)
* [ ] Veykril (optional)

programmerjake commented on text/0000-Project-Goals-2025h1.md:
```diff
@@ -0,0 +1,662 @@
+    * prepare to support asynchronous (and synchronous) generators in the language.
+* Continue helping Rust support low-level projects by [**stabilizing compiler options and tooling used by the Rust-for-Linux project**](https://rust-lang.github.io/rust-project-goals/2025h1/./rfl.html). In 2025H1 we plan to:
+    * implement [RFC #3716](https://github.com/rust-lang/rfcs/pull/3716) to allow stabilizing ABI-modifying compiler flags to control code generation, sanitizer integration, and so forth;
+    * taking the first step towards stabilizing [`build-std`](https://rust-lang.github.io/rust-project-goals/2025h1/https://doc.rust-lang.org/cargo/reference/unstable.html#build-std) by [creating a stable way to rebuild core with specific compiler options](./build-std.html);
```
links are still broken here...the latest commit didn't fix them all but you marked the threads as resolved

nikomatsakis commented on text/0000-Project-Goals-2025h1.md:
```diff
@@ -0,0 +1,662 @@
+    * implement [RFC #3716](https://github.com/rust-lang/rfcs/pull/3716) to allow stabilizing ABI-modifying compiler flags to control code generation, sanitizer integration, and so forth;
+    * taking the first step towards stabilizing [`build-std`](https://doc.rust-lang.org/cargo/reference/unstable.html#build-std) by [creating a stable way to rebuild core with specific compiler options](https://rust-lang.github.io/rust-project-goals/2025h1/./build-std.html);
+    * add rustdoc features to extract and customize rustdoc tests (`--extract-doctests`);
+    * stabilize clippy configuration like `.clippy.toml` and `CLIPPY_CONF_DIR`;
```
In general my take is that we should avoid making our stability guarantees overly complex. But we can weaken the section on clippy in particular if you like. I feel like the ability to accept this file without dying is all that needs to be stable; it's ok if there is deprecation and flags stop having the intended effect or what have you.

Manishearth replied:
Yes, that's the goal. I'm also fine with that being informally documented in this RFC thread for now.

RalfJung commented on text/0000-Project-Goals-2025h1.md:
```diff
@@ -0,0 +1,662 @@
+| [Model coherence in a-mir-formality](https://rust-lang.github.io/rust-project-goals/2025h1/formality.html)                                                          | [Niko Matsakis][]    | [types]                                             |
+| [Next-generation trait solver](https://rust-lang.github.io/rust-project-goals/2025h1/next-solver.html)                                                              | [lcnr][]            | [types]                                             |
+| [Nightly support for ergonomic SIMD multiversioning](https://rust-lang.github.io/rust-project-goals/2025h1/simd-multiversioning.html)                               | [Luca Versari][]        | [lang]                                              |
+| [Null and enum-discriminant runtime checks in debug builds](https://rust-lang.github.io/rust-project-goals/2025h1/null-enum-discriminant-debug-checks.html)         | [Bastian Kersting][]          | [compiler], [lang], [opsem]                         |
```
Where is the right place to ask clarifying questions about a goal? Seems a bit messy to do it all in this PR, but since I can't find a better place...

It is a bit unclear to me what is asked from us here in terms of a "design meeting". Like, what is there to design, opsem-wise? The opsem rules are fairly clear for this.

1c3t3a replied:
Yeah I figured the design meeting is not needed already. I can just remove that if wanted.

nikomatsakis replied:
I also wondered whether that meeting would be needed.

eholk commented:
@rfcbot reviewed

lcnr commented:
> https://rust-lang.github.io/rust-project-goals/2025h1/async.html#return-type-notation

stabilization decision should involve Types

> https://rust-lang.github.io/rust-project-goals/2025h1/async.html#unsafe-binders

same with any decision made here

davidtwco commented:
t-compiler are discussing our review capacity in our private Zulip stream prior to checking our boxes :)

CleanCut commented on text/0000-Project-Goals-2025h1.md:
```diff
@@ -0,0 +1,662 @@
+| [Externally Implementable Items](https://rust-lang.github.io/rust-project-goals/2025h1/eii.html)                                                                    | [Mara Bos][]         | [compiler], [lang]                                  |
+| [Field Projections](https://rust-lang.github.io/rust-project-goals/2025h1/field-projections.html)                                                                   | [Benno Lossin][]         | [compiler], [lang]                                  |
+| [Finish the libtest json output experiment](https://rust-lang.github.io/rust-project-goals/2025h1/libtest-json.html)                                                | [Ed Page][]           | [testing-devex]                                     |
+| [Implement Open API Namespace Support](https://rust-lang.github.io/rust-project-goals/2025h1/open-namespaces.html)                                                  | ![Help Wanted][] | [cargo], [compiler]                                 |
```
Reword to avoid confusing with [OpenAPI](https://www.openapis.org/)?

```suggestion
| [Implement Open Namespace Support on crates.io](https://rust-lang.github.io/rust-project-goals/2025h1/open-namespaces.html)                                                  | ![Help Wanted][] | [cargo], [compiler]                                 |
```

Turbo87 replied:
so it's not just me that was confused by this... 😅 

I agree that calling it "Open API Namespace Support" is a bit misleading. The original RFC called it "Packages as (optional) namespaces" and IMHO it would probably be best to keep that name to avoid such confusion.

nikomatsakis replied:
if somebody wants to open a PR against https://github.com/rust-lang/rust-project-goals/ I'd merge it. =) 

cc @epage

epage replied:
The name change is very intentional and the suggested fix is incorrect. It's a little clearer without title case ("open API namespaces")

This is a feature for open namespaces for Rust code with publish controls on crates.io. The naming is trying to make it clear that this is not registry namespaces which it is commonly confused with and which the RFC said it wasn't solving iirc.

Turbo87 replied:
idk if this is the right place to discuss this but it feels a bit strange to me that this was unilaterally renamed without discussion across the involved teams. "open namespaces" vs "registry namespaces" apparently makes a big difference to you but I assume for a regular user without a lot of context they sound basically equivalent.

epage replied:
> idk if this is the right place to discuss this but it feels a bit strange to me that this was unilaterally renamed without discussion across the involved teams. 

Before we get to into the weeds about that, a key distinction independent of the RFC is that this project goal is written for compiler and Cargo support.  At that level, there is no registry behavior and this is only a language feature of "open namespaces".

The initial versions of the RFC were unclear in their intent, being about [a single piece of software](https://github.com/rust-lang/rfcs/pull/3243#issuecomment-1064326145) and not organizational namespacing for registries.  As I better understood that, I reframed the RFC around this being a language feature [over several PRs](https://github.com/Manishearth/rfcs/pulls?q=is%3Apr+is%3Aclosed+author%3Aepage).  I did not rename the RFC itself.  I don't remember why, whether it was easier to be consistent on names (I've seen this with other RFCs that go through major revisions) or that I hadn't realized the level of confusion the name might cause until further discussion when it was too late.  The RFC does not speak at all about "registry namespacing", explicitly says its for [projects and not organizations](https://rust-lang.github.io/rfcs/3243-packages-as-optional-namespaces.html#user--org-namespaces), and in all of my communication about this, I've tried to make it very clear that this is not a general registry namespacing feature and that if people use it as such, they are likely to run into impedance mismatches.  Due to the connotations around "registry namespacing" (being organizational focused), I would go as far as saying that calling this "registry namespacing" isn't just a misnomer (as the focus is on the language feature) but goes against the intent of the RFC (oriented around APIs, not organizations).

>  "open namespaces" vs "registry namespaces" apparently makes a big difference to you but I assume for a regular user without a lot of context they sound basically equivalent.

The first term describes a language feature.  The second term describes a registry feature.  How people will understand this is the context it is communicated in which is up to us.

davidtwco commented:
> t-compiler are discussing our review capacity in our private Zulip stream prior to checking our boxes :)

We've agreed who will take point in reviewing each of the goals that we think will need a dedicated reviewer, so checking both Wesley and I's boxes.

pthariensflame commented on text/0000-Project-Goals-2025h1.md:
```diff
@@ -0,0 +1,662 @@
+
+## Summary
+
+Propose a slate of 40 project goals for 2025H1, including 3 flagship goals:
```
40, or 41?  Since it’s 41 below.

jdno commented on text/0000-Project-Goals-2025h1.md:
```diff
@@ -0,0 +1,662 @@
+| [Rust Vision Document](https://rust-lang.github.io/rust-project-goals/2025h1/rust-vision-doc.html)                                                                  | [Niko Matsakis][]    | [leadership-council]                                |
+| [SVE and SME on AArch64](https://rust-lang.github.io/rust-project-goals/2025h1/arm-sve-sme.html)                                                                    | [David Wood][]       | [compiler], [lang], [types]                         |
+| [Scalable Polonius support on nightly](https://rust-lang.github.io/rust-project-goals/2025h1/Polonius.html)                                                         | [Rémy Rakic][]             | [types]                                             |
+| [Secure quorum-based cryptographic verification and mirroring for crates.io](https://rust-lang.github.io/rust-project-goals/2025h1/verification-and-mirroring.html) | [@walterhpearce][]   | [cargo], [crates-io], [infra], [leadership-council] |
```
T-Infra discussed this in its [weekly team meeting](https://rust-lang.zulipchat.com/#narrow/channel/242791-t-infra/topic/meeting.202025-01-27) today, and has a similar (non-blocking) concern around the status of the signing RFC and the likelihood of it being implemented in H1.

nikomatsakis replied:
I'll take a look. I have tried to ensure that the goals are framed as "a decision is required" from the team but not that the decision is *positive*, though I think if the team is skeptical of the idea, that is a good thing to know up front.

nikomatsakis replied:
I've incorporated the edits intended to resolve this point.

jdno reviewed (commented):
Given that we're a month into H1 and the RFC still hasn't been ratified, I think it'd be good to do a proper retrospective of 2024H2 with all stakeholders to figure out how to improve the process. Would be great to hit the ground running in 2025H2. 🙂

nikomatsakis commented:
> Given that we're a month into H1 and the RFC still hasn't been ratified, I think it'd be good to do a proper retrospective of 2024H2 with all stakeholders to figure out how to improve the process. Would be great to hit the ground running in 2025H2. 🙂

One is planned, yeah.  It's the first hand-over and we've got some kinks to work out.

nikomatsakis commented on text/0000-Project-Goals-2025h1.md:
```diff
@@ -0,0 +1,662 @@
+
+Project goals can vary in scope from an internal refactoring that affects only one team to a larger cross-cutting initiative. No matter its scope, accepting a goal should never be interpreted as a promise that the team will make any future decision (e.g., accepting an RFC that has yet to be written). Rather, it is a promise that the team are aligned on the contents of the goal thus far (including ...
+
+Of the proposed goals, a small subset are selected by the roadmap owner as **flagship goals**. Flagship goals are chosen for their high impact (many Rust users will be impacted) and their shovel-ready nature (the org is well-aligned around a concrete plan). Flagship goals are the ones that will feature most prominently in our public messaging and which should be prioritized by Rust teams where ne ...
```
This was launched via [RFC 3614](https://rust-lang.github.io/rfcs/3614-project-goals.html); this is the second round of experimentation, as discussed on Zulip [here](https://rust-lang.zulipchat.com/#narrow/channel/392734-council/topic/Project.20Goals.202025H1/near/478330305). There is an [open PR](https://github.com/rust-lang/team/pull/1641) to create a "landing pad" team (with [this charter](https://rust-lang.github.io/rust-project-goals/admin/team.html documents the team charter)) -- presuming the program is made permanent via an RFC, I would expect that team to move somewhere else, though precisely *where* [has been a topic of active discussion](https://rust-lang.zulipchat.com/#narrow/channel/392734-council/topic/team.20for.20the.20goals.20program/near/484805177).

nikomatsakis replied:
(The lead of the team is the roadmap owner—currently me—as described in the charter.)

jdno replied:
Feel free to resolve the conversation. We can continue the conversation in the other channels (or the future RFC) if necessary...

JoelMarcey commented on text/0000-Project-Goals-2025h1.md:
```diff
@@ -0,0 +1,662 @@
+| [Publish first version of StableMIR on crates.io](https://rust-lang.github.io/rust-project-goals/2025h1/stable-mir.html)                                            | [Celina G. Val][]        | [compiler], [project-stable-mir]                    |
+| [Research: How to achieve safety when linking separately compiled code](https://rust-lang.github.io/rust-project-goals/2025h1/safe-linking.html)                    | [Mara Bos][]         | [compiler], [lang]                                  |
+| [Run the 2025H1 project goal program](https://rust-lang.github.io/rust-project-goals/2025h1/stabilize-project-goal-program.html)                                    | [Niko Matsakis][]    | [leadership-council]                                |
+| [Rust Specification Testing](https://rust-lang.github.io/rust-project-goals/2025h1/spec-testing.html)                                                               | [Connor Horman][]     | [bootstrap], [compiler], [lang], [spec]             |
```
I believe we are removing this goal, right?

JoelMarcey commented on text/0000-Project-Goals-2025h1.md:
```diff
@@ -0,0 +1,662 @@
+| **Discussion and moral support**                                                                                                  |              |                                                                                                |
+| ↳ [Making compiletest more maintainable: reworking directive handling](https://rust-lang.github.io/rust-project-goals/2025h1/compiletest-directive-rework.html#ownership-and-team-asks)   | [Jieyou Xu][]    | including consultations for desired test behaviors and testing infra consumers                 |
+| **RFC decision**                                                                                                                  |              |                                                                                                |
+| ↳ [Rust Specification Testing](https://rust-lang.github.io/rust-project-goals/2025h1/spec-testing.html#ownership-and-team-asks)                                                           | [Connor Horman][] |                                                                                                |
```
Remove

JoelMarcey commented on text/0000-Project-Goals-2025h1.md:
```diff
@@ -0,0 +1,662 @@
+| ↳ [rustc-perf improvements](https://rust-lang.github.io/rust-project-goals/2025h1/perf-improvements.html#ownership-and-team-asks)                                                         | [David Wood][]    | Update performance regression policy                                                           |
+| **RFC decision**                                                                                                                  |               |                                                                                                |
+| ↳ [ABI-modifying compiler flags](https://rust-lang.github.io/rust-project-goals/2025h1/rfl.html#ownership-and-team-asks)                                                                  | [Niko Matsakis][] | [RFC #3716](https://github.com/rust-lang/rfcs/pull/3716), currently in PFCP                                                                 |
+| ↳ [Rust Specification Testing](https://rust-lang.github.io/rust-project-goals/2025h1/spec-testing.html#ownership-and-team-asks)                                                           | [Connor Horman][]  |                                                                                                |
```
Remove

JoelMarcey commented on text/0000-Project-Goals-2025h1.md:
```diff
@@ -0,0 +1,662 @@
+| ↳ [Land nightly experiment for SVE types](https://rust-lang.github.io/rust-project-goals/2025h1/arm-sve-sme.html#ownership-and-team-asks)                                               | [David Wood][]    |                                                                                                                 |
+| ↳ [Extending type system to support scalable vectors](https://rust-lang.github.io/rust-project-goals/2025h1/arm-sve-sme.html#ownership-and-team-asks)                                   | [David Wood][]    |                                                                                                                 |
+| ↳ [Field Projections](https://rust-lang.github.io/rust-project-goals/2025h1/field-projections.html#ownership-and-team-asks)                                                             | [Benno Lossin][]      |                                                                                                                 |
+| ↳ [Rust Specification Testing](https://rust-lang.github.io/rust-project-goals/2025h1/spec-testing.html#ownership-and-team-asks)                                                         | [Connor Horman][]  |                                                                                                                 |
```
Remove

JoelMarcey commented on text/0000-Project-Goals-2025h1.md:
```diff
@@ -0,0 +1,662 @@
+| ↳ [Integration of the FLS into the Rust Project](https://rust-lang.github.io/rust-project-goals/2025h1/spec-fls-integration.html#ownership-and-team-asks)   | [Joel Marcey][]  | Take ownership of the FLS (prior to, or shortly into January 2025). |
+| **RFC decision**                                                                                    |              |                                                                     |
+| ↳ [Integrate FLS into T-spec processes](https://rust-lang.github.io/rust-project-goals/2025h1/spec-fls-integration.html#ownership-and-team-asks)            | [Joel Marcey][]  | End of March 2025                                                   |
+| ↳ [Rust Specification Testing](https://rust-lang.github.io/rust-project-goals/2025h1/spec-testing.html#ownership-and-team-asks)                             | [Connor Horman][] |                                                                     |
```
Remove

JoelMarcey commented on text/0000-Project-Goals-2025h1.md:
May not be needed any longer given we are removing the specification testing goal.

JoelMarcey reviewed (commented):
Approved with the suggestions I made.

spastorino commented on text/0000-Project-Goals-2025h1.md:
```diff
@@ -0,0 +1,662 @@
+| [Implement restrictions, prepare for stabilization](https://rust-lang.github.io/rust-project-goals/2025h1/restrictions.html)                                        | [Jacob Pratt][]         | [compiler], [lang]                                  |
+| [Improve state machine codegen](https://rust-lang.github.io/rust-project-goals/2025h1/improve-rustc-codegen.html)                                                   | [Folkert de Vries][]      | [compiler], [lang]                                  |
+| [Instrument the Rust standard library with safety contracts](https://rust-lang.github.io/rust-project-goals/2025h1/std-contracts.html)                              | [Celina G. Val][]        | [compiler], [libs]                                  |
+| [Integration of the FLS into the Rust Project](https://rust-lang.github.io/rust-project-goals/2025h1/spec-fls-integration.html)                                     | [Joel Marcey][]      | [release], [spec]                                   |
```
```suggestion
| [Integration of the FLS into the Rust Project](https://rust-lang.github.io/rust-project-goals/2025h1/spec-fls-publish.html)                                     | [Joel Marcey][]      | [release], [spec]                                   |
```

Mark-Simulacrum replied:
re:release ask here, it's not clear to me that release is the right team. @JoelMarcey, perhaps we can sync offline or otherwise discuss what you're looking for? Release generally doesn't own the "artifact pipeline", only the public-facing bits of it (i.e., kicking it off, attaching release notes, blog post). If we need another book published that's probably more of a T-infra/bootstrap ask, or perhaps just not an ask at all (if it works the same as every other book ~anyone can land a PR adding one).

JoelMarcey replied:
@Mark-Simulacrum Just [commented in Zulip](https://rust-lang.zulipchat.com/#narrow/channel/399173-t-spec/topic/Spec.20infra) as well. I had thought the `t-spec` team believed `t-release` may be necessary for tooling integration that could be needed for FLS publishing. If that's not true, I can have a conversation with `t-spec` at our meeting tomorrow.

nikomatsakis replied:
During our sync meeting today we determined that the only ask needed here was "standard reviews" from the bootstrap team (and we did not expect any particular load or challenge in doing those reviews). @JoelMarcey to update the goal. Resolving this thread.

JoelMarcey replied:
https://github.com/rust-lang/rust-project-goals/pull/241

JoelMarcey replied:
Do I need to make any updates to this RFC?

JoelMarcey commented on text/0000-Project-Goals-2025h1.md:
```diff
@@ -0,0 +1,662 @@
+| Goal                                                                                      | Point of contact | Notes          |
+| ---                                                                                       | ---         | ---            |
+| **Discussion and moral support**                                                          |             |                |
+| ↳ [Integrate FLS into release process](https://rust-lang.github.io/rust-project-goals/2025h1/spec-fls-integration.html#ownership-and-team-asks)   | [Joel Marcey][] | Februrary 2025 |
```
```suggestion
| ↳ [Integrate FLS into release process](https://rust-lang.github.io/rust-project-goals/2025h1/spec-fls-publish.html#ownership-and-team-asks)   | [Joel Marcey][] | Februrary 2025 |
```

JoelMarcey commented on text/0000-Project-Goals-2025h1.md:
```diff
@@ -0,0 +1,662 @@
+| **Discussion and moral support**                                                          |             |                |
+| ↳ [Integrate FLS into release process](https://rust-lang.github.io/rust-project-goals/2025h1/spec-fls-integration.html#ownership-and-team-asks)   | [Joel Marcey][] | Februrary 2025 |
+| **Standard reviews**                                                                      |             |                |
+| ↳ [Integrate FLS into release process](https://rust-lang.github.io/rust-project-goals/2025h1/spec-fls-integration.html#ownership-and-team-asks)   | [Joel Marcey][] |                |
```
```suggestion
| ↳ [Integrate FLS into release process](https://rust-lang.github.io/rust-project-goals/2025h1/spec-fls-publish.html#ownership-and-team-asks)   | [Joel Marcey][] |                |
```

JoelMarcey commented on text/0000-Project-Goals-2025h1.md:
```diff
@@ -0,0 +1,662 @@
+| Goal                                                                                                | Point of contact | Notes                                                               |
+| ---                                                                                                 | ---          | ---                                                                 |
+| **Miscellaneous**                                                                                   |              |                                                                     |
+| ↳ [Integration of the FLS into the Rust Project](https://rust-lang.github.io/rust-project-goals/2025h1/spec-fls-integration.html#ownership-and-team-asks)   | [Joel Marcey][]  | Take ownership of the FLS (prior to, or shortly into January 2025). |
```
```suggestion
| ↳ [Integration of the FLS into the Rust Project](https://rust-lang.github.io/rust-project-goals/2025h1/spec-fls-publish.html#ownership-and-team-asks)   | [Joel Marcey][]  | Take ownership of the FLS (prior to, or shortly into January 2025). |
```

JoelMarcey commented on text/0000-Project-Goals-2025h1.md:
```diff
@@ -0,0 +1,662 @@
+| **Miscellaneous**                                                                                   |              |                                                                     |
+| ↳ [Integration of the FLS into the Rust Project](https://rust-lang.github.io/rust-project-goals/2025h1/spec-fls-integration.html#ownership-and-team-asks)   | [Joel Marcey][]  | Take ownership of the FLS (prior to, or shortly into January 2025). |
+| **RFC decision**                                                                                    |              |                                                                     |
+| ↳ [Integrate FLS into T-spec processes](https://rust-lang.github.io/rust-project-goals/2025h1/spec-fls-integration.html#ownership-and-team-asks)            | [Joel Marcey][]  | End of March 2025                                                   |
```
```suggestion
| ↳ [Integrate FLS into T-spec processes](https://rust-lang.github.io/rust-project-goals/2025h1/spec-fls-publish.html#ownership-and-team-asks)            | [Joel Marcey][]  | End of March 2025                                                   |
```

JoelMarcey reviewed (commented):
A few link changes needed.

celinval reviewed (approved):
It looks good to me.

Ps.: The checkbox doesn't seem to work for me. I can click the box, but github doesn't save it.

nikomatsakis commented:
Hi folks,

A few things--- 

1. I pushed another round of updates from the rust-project-goals repository
2. We've started pinging team leads -- the pinging will continue until boxes are checked (or reasons are given!)

One thing to note on point 2 is that I did not do the best job of pre-circulating these goals with all relevant parties. So if this is the first time you are seeing an ask of your team, please feel free to ask questions. Now that we have a project goals team that is forming and can share the load, I expect we'll do a better job about circulating goals up front next time.

nikomatsakis commented:
> > https://rust-lang.github.io/rust-project-goals/2025h1/async.html#return-type-notation
> 
> stabilization decision should involve Types
> 
> > https://rust-lang.github.io/rust-project-goals/2025h1/async.html#unsafe-binders
> 
> same with any decision made here

good catch, let me check that @lcnr

nikomatsakis commented:
Update: After some conversation with release team, we hammered out the requirements for spec goal, and that team is removed from the goal. Therefore I've removed them from the RFC checklist, as this was the only ask.

I've also taken the liberty of checking the box of @calebcartwright on behalf of the ask for "discussion and moral support" for the "testing-devex" working group, given that (a) they've not responded to pings on Zulip, so they are probably away; (b) the ask is just for basic approval and (c) I'm not even sure that working groups need to check boxes (though I don't see how it hurts).

nikomatsakis commented:
I believe at this point the only outstanding issue is hammering out agreement about the [Secure quorum-based cryptographic verification and mirroring for crates.io](https://rust-lang.github.io/rust-project-goals/2025h1/verification-and-mirroring.html) goal, and we have ongoing discussion on Zulip about that. Here's hoping we can this into FCP tomorrow!

tmandry commented on text/0000-Project-Goals-2025h1.md:
```diff
@@ -0,0 +1,661 @@
+
+#### Why these particular flagship goals?
+
+[**Async.**](https://rust-lang.github.io/rust-project-goals/2025h1/./async.html) Rust is a great fit for server development thanks to its ability to scale to very high load while retaining low memory usage and tight tail latency. 52% of the respondents in the [2023 Rust survey](https://blog.rust-lang.org/2024/02/19/2023-Rust-Annual-Survey-2023-results.html) indicated that they use Rust to build s ...
```
```suggestion
[**Async.**](https://rust-lang.github.io/rust-project-goals/2025h1/./async.html) Rust is a great fit for server development thanks to its ability to scale to very high load while retaining low memory usage and tight tail latency. 52% of the respondents in the [2023 Rust survey](https://blog.rust-lang.org/2024/02/19/2023-Rust-Annual-Survey-2023-results.html) indicated that they use Rust to build server-side or backend applications. In [2025H1 our plan](https://rust-lang.github.io/rust-project-goals/2025h1/./async.html) is to deliver (a) improved support for async-fn-in-traits, completely subsuming the functionality of the [`async-trait` crate](https://crates.io/crates/async-trait); (b) design progress for sync and async generators, simplifying the creation of iterators and async data streams; (c) and improve the ergonomics of `Pin`, making lower-level async coding more approachable. These items together start to unblock the creation of the next generation of async libraries in the wider ecosystem, as progress there has been blocked on a stable solution for async traits and streams.
```

The project goal text doesn't promise a final design, certainly for both sync and async, though we hope to have an RFC for sync generators.

tmandry replied:
Since Github is not highlighting the diff, it was: replace `finalize a design` with `design progress`.

nikomatsakis commented:
Hi all, 

I've incorporated the latest round of edits. At this point the only remaining sign-off required is from the cargo team (@ehuss as lead).

tmandry commented:
I've checked my box, but I want to note a concern: There are substantially _more_ accepted goals in this round than in the previous, and currently only one under "Goals not accepted". It doesn't seem like prioritization was as carefully done this round, and as a team lead I was never explicitly asked to do this. If that continues to be the norm I worry about project goals becoming a process-heavy crystallization of a problem the Project has struggled with for years: Trying to focus on too many things with too little time.

What's the right number of project goals? I don't claim to know, but if I can't easily remember all the goals with asks for my team I think we are already flirting with it being too high. A big advantage of project goals is that it allows us to set aside dedicated time for higher-level priorities that we actually intend to follow through on. If instead it becomes a blessed way to bypass more granular prioritization decisions for things that _aren't_ that, it will make the project less efficient.

Granted, another benefit is that it allows goal owners to get a quick "vibe check" from the team. But I think we have to be careful about separating this from a commitment to actually dedicate future resources to a problem. There are some goals that had meaningful early feedback from the lang team during this process – that's a win!

For others it felt more to me like "I see no reason why you _shouldn't_ do this and since you don't seem to be asking for that much, fine". My gut tells me we should not accept goals like this, since that can easily lead to a mismatch in expectations. There should always be a good understanding of why we _should_ accept a goal, according to some criteria, instead of an absence of reasons _not_ to. Keeping in mind that project goals are not intended to be the only entry way to the project – rejecting a goal does not mean rejecting the effort altogether.

nikomatsakis commented:
@tmandry Solid concerns. I also have wondered if we have too many goals. At the same time, I think that as a project we've done a better job of calculating and estimating resources for many teams than we did before (e.g., the compiler team did explicitly discuss review bandwidth, and the spec team used the goals to anticipate what kind of edits to the reference would be required).

When it comes to the lang team specifically, I think we probably need to think more about how to better integrate goal review and prioritization into our processes. It seems like a good task for the goals team to work with teams to figure out how to do the same for next time. Something we could do better I think is taking a skeptical look at each goal to determine if it is realistically accounting for the resources required and if we can really support it at this time.

tmandry commented:
> When it comes to the lang team specifically, I think we probably need to think more about how to better integrate goal review and prioritization into our processes.

Yes, and since we are working to revamp some of those processes anyway it will be a good time to do it. Part of my reaction here is that my perspective on e.g. _how much commitment an experiment requires from us_ has changed in the last month or two since we originally reviewed the goal slate. But I'm okay with these lessons taking time to trickle down into the process, as long as they eventually do.

> Something we could do better I think is taking a skeptical look at each goal to determine if it is realistically accounting for the resources required and if we can really support it at this time.

That's exactly what I want.

calebcartwright reviewed (approved):
Adding my approval just for the sake posterity to indicate my support/agreement with my box having been checked - https://github.com/rust-lang/rfcs/pull/3764#issuecomment-2625704385

thanks @nikomatsakis !

nikomatsakis commented:
Updates:

1. I'm going to publish a new version soon that removes the "RFC decision" ask that is blocked cargo/crates-io signoff.
2. Wearing my lang team co-lead hat, @tmandry and I talked and decided to pare back some of the lang team asks that we are willing to support. I'll be pushing updates on that too -- nothing too major, in some cases we've decided to limit things to experiments because we don't think we can manage to reach an RFC decision, etc. (We're going to look into the best process for this next time.)

I am hoping to move this RFC to FCP by end of day.

nikomatsakis commented:
I've updated the goal. The lang goals are a bit more limited and @ehuss the "RFC decision" has been removed from the mirror/quorum goal, per your request. The formatting has also changed to a new style I think helps make clearer what's going on, where there is a table with one row per goal.

nikomatsakis commented:
**Huzzah!** From what I can see, all boxes are checked, so I am marking this RFC as entering **final comment period**.

**FCP will officially end Monday Feb 18, but I intend to start moving forward with goal setup immediately, given that this is mostly a matter of creating issues and the rest.**

@rustbot labels +final-comment-period -proposed-final-comment-period

nikomatsakis commented:
**Huzzah!** This RFC is officially accepted. Tracking issues for the various goals have been created on the rust-lang/rust-project-goals repository, so you can track progress there on the [2025h1 milestone](https://github.com/rust-lang/rust-project-goals/milestone/3).
